scripts/deleteSnapsets.js

```diff
@@ -0,0 +1,162 @@
+dbLoader.prefsSafesViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findSnapsetPrefsSafes";
+dbLoader.gpiiKeyViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findGpiiKeysByPrefsSafeId";
+dbLoader.parsedCouchDbUrl = url.parse(dbLoader.couchDbUrl);
+dbLoader.docsToRemove=[];
+
+/**
+ * Find the Prefs Safes of type "snapSet", add them to an array of records to
+ * remove, then use them to find their associated GPII Key(s).
+ * @param {Object} response - The response from the data base containing the
+ *                            snapSet Prefs Safes as a JSON string.
+ */
+dbLoader.processSnapsets = function (response) {
+    var snapSetsString = "";
+    response.setEncoding("utf8");
+    response.on("data", function (chunk) {
+        snapSetsString += chunk;
+    });
+    response.on("end", function () {
+        dbLoader.snapSets = JSON.parse(snapSetsString);
+        fluid.each(dbLoader.snapSets.rows, function (aSnapset) {
+            aSnapset.value._deleted = true;
+            dbLoader.docsToRemove.push(aSnapset.value);
+        });
+        dbLoader.addGpiiKeysAndBulkDelete(dbLoader.snapSets.rows, dbLoader.docsToRemove);
+    });
+};
+
+/**
+ * Find the GPII key records associated with the given snapset Prefs Safes,
+ * delete each one as found, and then batch delete all the snapset Prefs Safes.
+ * @param {Array} snapSets      - The snapsets of interest.
+ * @param {Array} docsToRemove  - Array of records to delete.
+ */
+dbLoader.addGpiiKeysAndBulkDelete = function (snapSets, docsToRemove) {
+    fluid.each(snapSets, function (aSnapset) {
+        var gpiiKeyId = aSnapset.value._id;
+        fluid.log("Snapset: " + gpiiKeyId);
+        var gpiiKeyViewUrl = dbLoader.gpiiKeyViewUrl + "?key=%22" + gpiiKeyId + "%22";
+        var getGpiiKeysRequest = http.request(gpiiKeyViewUrl, function (resp) {
+            var respString = "";
+            resp.setEncoding("utf8");
+            resp.on("data", function (chunk) {
+                respString += chunk;
+            });
+            // This response "end" event doesn't finish until after the call to
+            // doBatchDelete() below, which is too late to add the GPII Key to
+            // docsToRemove and have them removed.
+            // Note - if there is a way to determine the final "end" event, then
+            // it could  call doBatchDelete().
+            resp.on("end", function () {
+                var gpiiKeyRecords = JSON.parse(respString);
+                fluid.each(gpiiKeyRecords.rows, function (gpiiKey) {
+                    fluid.log("GPII Key: " + gpiiKey.value._id);
+                    dbLoader.deleteGpiiKey(gpiiKey.value);
+                });
+            });
+        });
+        getGpiiKeysRequest.on("error", function (e) {
+            fluid.log("Error finding snapsets' associated GPII Keys: " + e.message);
+        });
+        getGpiiKeysRequest.end();
+    });
+    dbLoader.doBatchDelete(docsToRemove);
```

Comment:
Deleting all prefs safe records here at the end of the script doesn't guarantee it to be run as the last task due to the aync nature of http requests above for getting and deleting gpii keys.

Reply:
Right, but it's another instance of "overtaken by events":  The latest code handles the asynchrony.